### PR TITLE
Add Whisper static generation

### PR DESCRIPTION
**Pull request by Spycsh**
# What does this PR do?

Add static KV cache for Whisper family models. Previously it takes huge amount of time to do the Whisper inference (each first inference for each different input audio) on HPU because of dynamic shape of KV cache in SDPA attention module. With this PR the latency should be drastically reduced on HPU.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by Spycsh** [comment]
```
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import contextlib
import os
import time
import urllib.request

import numpy as np
import torch
from datasets import Audio, Dataset
from pydub import AudioSegment


class WhisperModel:
    """Convert audio to text."""

    def __init__(self, model_name_or_path="openai/whisper-small", language="english", device="cpu", hpu_max_len=8192):
        if device == "hpu":
            # Explicitly link HPU with Torch
            from optimum.habana.transformers.modeling_utils import adapt_transformers_to_gaudi

            adapt_transformers_to_gaudi()
        from transformers import WhisperForConditionalGeneration, WhisperProcessor

        self.device = device
        asr_model_name_or_path = os.environ.get("ASR_MODEL_PATH", model_name_or_path)
        print("Downloading model: {}".format(asr_model_name_or_path))
        self.model = WhisperForConditionalGeneration.from_pretrained(asr_model_name_or_path).to(self.device)
        self.processor = WhisperProcessor.from_pretrained(asr_model_name_or_path)
        self.model.eval()

        self.hpu_max_len = hpu_max_len
        self.language = language

        if device == "hpu":
            #self._warmup_whisper_hpu_graph("https://paddlespeech.bj.bcebos.com/Parakeet/docs/demos/labixiaoxin.wav")
            #self._warmup_whisper_hpu_graph("https://github.com/Spycsh/assets/raw/main/ljspeech_60s_audio.wav")
            #self._warmup_whisper_hpu_graph("https://github.com/Spycsh/assets/raw/main/ljspeech_30s_audio.wav")
            pass

    def _audiosegment_to_librosawav(self, audiosegment):
        # https://github.com/jiaaro/pydub/blob/master/API.markdown#audiosegmentget_array_of_samples
        # This way is faster than librosa.load or HuggingFace Dataset wrapper
        channel_sounds = audiosegment.split_to_mono()[:1]  # only select the first channel
        samples = [s.get_array_of_samples() for s in channel_sounds]

        fp_arr = np.array(samples).T.astype(np.float32)
        fp_arr /= np.iinfo(samples[0].typecode).max
        fp_arr = fp_arr.reshape(-1)

        return fp_arr

    def _warmup_whisper_hpu_graph(self, url):
        print("[ASR] fetch warmup audio...")
        urllib.request.urlretrieve(
            url,
            "warmup.wav",
        )
        print("[ASR] warmup...")
        waveform = AudioSegment.from_file("warmup.wav").set_frame_rate(16000)
        waveform = self._audiosegment_to_librosawav(waveform)

        try:
            processed_inputs = self.processor(
                waveform,
                return_tensors="pt",
                truncation=False,
                padding="longest",
                return_attention_mask=True,
                sampling_rate=16000,
            )
        except RuntimeError as e:
            if "Padding size should be less than" in str(e):
                # short-form
                processed_inputs = self.processor(
                    waveform,
                    return_tensors="pt",
                    sampling_rate=16000,
                )
            else:
                raise e

        if processed_inputs.input_features.shape[-1] < 3000:
            # short-form
            processed_inputs = self.processor(
                waveform,
                return_tensors="pt",
                sampling_rate=16000,
            )
        else:
            processed_inputs["input_features"] = torch.nn.functional.pad(
                processed_inputs.input_features,
                (0, self.hpu_max_len - processed_inputs.input_features.size(-1)),
                value=-1.5,
            )
            processed_inputs["attention_mask"] = torch.nn.functional.pad(
                processed_inputs.attention_mask,
                (0, self.hpu_max_len + 1 - processed_inputs.attention_mask.size(-1)),
                value=0,
            )

        predicted_ids = self.model.generate(
            **(
                processed_inputs.to(
                    self.device,
                )
            ),
            language=self.language,
        )
        result = self.processor.tokenizer.batch_decode(predicted_ids, skip_special_tokens=True, normalize=True)[0]
        print(result)

    def audio2text(self, audio_path):
        """Convert audio to text.

        audio_path: the path to the input audio, e.g. ~/xxx.mp3
        """
        start = time.time()

        try:
            waveform = AudioSegment.from_file(audio_path).set_frame_rate(16000)
            waveform = self._audiosegment_to_librosawav(waveform)
        except Exception as e:
            print(f"[ASR] audiosegment to librosa wave fail: {e}")
            audio_dataset = Dataset.from_dict({"audio": [audio_path]}).cast_column("audio", Audio(sampling_rate=16000))
            waveform = audio_dataset[0]["audio"]["array"]

        try:
            processed_inputs = self.processor(
                waveform,
                return_tensors="pt",
                truncation=False,
                padding="longest",
                return_attention_mask=True,
                sampling_rate=16000,
            )
        except RuntimeError as e:
            if "Padding size should be less than" in str(e):
                # short-form
                processed_inputs = self.processor(
                    waveform,
                    return_tensors="pt",
                    sampling_rate=16000,
                )
            else:
                raise e
        if processed_inputs.input_features.shape[-1] < 3000:
            # short-form
            processed_inputs = self.processor(
                waveform,
                return_tensors="pt",
                sampling_rate=16000,
            )

        predicted_ids = self.model.generate(
            **(
                processed_inputs.to(
                    self.device,
                )
            ),
            language=self.language,
        )
        # pylint: disable=E1101
        result = self.processor.tokenizer.batch_decode(predicted_ids, skip_special_tokens=True, normalize=True)[0]
        if self.language in ["chinese", "mandarin"]:
            from zhconv import convert

            result = convert(result, "zh-cn")
        print(f"generated text in {time.time() - start} seconds, and the result is: {result}")
        return result


if __name__ == "__main__":
    asr = WhisperModel(model_name_or_path="openai/whisper-tiny", language="english", device="hpu")
    urllib.request.urlretrieve(f"https://github.com/Spycsh/assets/raw/main/ljspeech_60s_audio.wav", "sample.wav")
    text = asr.audio2text("sample.wav")

    # Test multilanguage asr
    urllib.request.urlretrieve(
        "https://paddlespeech.bj.bcebos.com/Parakeet/docs/demos/labixiaoxin.wav",
        "sample.wav",
    )
    asr.language = "chinese"
    text = asr.audio2text("sample.wav")

    asr.language = "english"
    urllib.request.urlretrieve(
        "https://github.com/intel/intel-extension-for-transformers/raw/main/intel_extension_for_transformers/neural_chat/assets/audio/sample.wav",
        "sample.wav",
    )
    text = asr.audio2text("sample.wav")

    for i in [5, 10, 30, 60]:
        urllib.request.urlretrieve(f"https://github.com/Spycsh/assets/raw/main/ljspeech_{i}s_audio.wav", "sample.wav")
        text = asr.audio2text("sample.wav")
```

**Comment by emascarenhas** [comment]
Hi @Spycsh ,

I haven't reviewed this code as yet and have few initial questions.

Please confirm code merges cleanly with OH main branch and runs with latest 1.17.1 as well as next 1.18.0 latest build.

How did you test these changes? Do you need to add any new tests to the CI? 
How does this fit in with the run_speech_recogntion_seq2seq.py script in examples/speech-recognition?

In addition, please run "tests/ci/fast_tests.sh" after installing ".[tests]" and any slow tests, e.g., the ones that run the whisper model in the the CI test suite although e.g., 
GAUDI2_CI=1 RUN_SLOW=1 python -m pytest 
 tests/test_examples.py -v -s -k whisper

Also do "pip install -U ruff; make style" and check for any issues.

**Comment by Spycsh** [comment]
Hi @emascarenhas , thanks for suggestions. I've added the test, passed style check and passed the test on habana 1.16.1 and OH merged with the latest main branch. I currently do not have environment with newer habana driver but I think it should work. 

Please tell me if you have further suggestions!

**Comment by emascarenhas** [comment]
I got this result
***** eval metrics *****
  eval_loss                   =     1.0118
  eval_model_preparation_time =     0.0243
  eval_runtime                = 0:11:54.28
  eval_samples                =       2894
  eval_samples_per_second     =      4.052
  eval_steps_per_second       =      0.127
  eval_wer                    =      3.024
  max_memory_allocated (GB)   =      53.45
  memory_allocated (GB)       =      53.44
  total_memory_available (GB) =      94.62

Do you need to adjust the performance criteria?

tests/test_speech_recognition_example.py:69: AssertionError
=================================================================== short test summary info ====================================================================
FAILED tests/test_speech_recognition_example.py::test_speech_recognition_bf16[openai/whisper-small-32-2.892] - assert 4.052 >= (2 * 2.892)
================================================================ 1 failed in 733.05s (0:12:13) =================================================================

**Comment by emascarenhas** [comment]
@Spycsh , 
How can I tell whether the new code in optimum/habana/transformers/models/whisper/modeling_whisper.py is used?
Since you are using the run_speech_recognition_seq2seq.py script from earlier without any change, does it invoke any of the new definitions/methods in the modeling_whisper.py file?

**Comment by Spycsh** [comment]
Hi @emascarenhas , on my gaudi2 the throughput (eval_samples_per_second) should be `7,757` vs. baseline `2.892` so I assert the PR throughtput should be >=2 baseline throughput. However your tested data show a `4.052` vs. baseline `2.892`. I have no idea why the baseline is close but the PR throughput vary a lot between our hardwares. We can still lower the value of multiple a little bit to make it pass in your environment (e.g. PR throughput >= 1.2 baseline throughput).

My detailed logs of the latest run are shown below:

```
***** eval metrics *****
  eval_loss                   =     1.0118
  eval_model_preparation_time =     0.0449
  eval_runtime                = 0:06:17.92
  eval_samples                =       2894
  eval_samples_per_second     =      7.658
  eval_steps_per_second       =      0.241
  eval_wer                    =     2.0325
  max_memory_allocated (GB)   =      13.16
  memory_allocated (GB)       =      13.15
  total_memory_available (GB) =      94.62
[INFO|modelcard.py:449] 2024-09-05 01:38:31,766 >> Dropping the following result as it does not have all the necessary fields:
{'task': {'name': 'Automatic Speech Recognition', 'type': 'automatic-speech-recognition'}, 'dataset': {'name': 'mozilla-foundation/common_voice_11_0 hi', 'type': 'mozilla-foundation/common_voice_11_0', 'config': 'hi', 'split': 'test', 'args': 'hi'}}
```

My command to trigger the test:

```
pytest -v -s tests/test_speech_recognition_example.py
```

Regarding to your another question about where is `modeling_whisper.py` used:

Previously `run_speech_recognition_seq2seq.py` already used `openai/whisper-small` to measure the throughput, do the finetuning on `hindi` language. It use the native transformers implementation for Whisper and do not provide any optimization. Meanwhile, it has a super slow warmup speed because it does not have a static KV cache shape optimization. Every time decoding iteration has a increasing input length so multiple different shape of HPU graphs will be initialized and every initialization takes some time so it would be very slow. My solution is to keep the KV shape the same padded to max-length during decoding so HPU graph will be as few as possible. That is what this PR aim to provide.

After this PR, the `run_speech_recognition_seq2seq.py` will not go into native transformers implementation of Whisper but go into the attention logics I write in `modeling_whisper.py`. And this should benefit from static generation.

Welcome to further questions!

**Comment by emascarenhas** [comment]
It makes sense.

I suppose we don't need to have tests/test_speech_recognition_example.py because we already have an examples test in the CI that invokes run_speech_recognition_seq2seq.py. We don't need to run the same test run twice. If it is running the same test with exact same code coverage, then I think we should remove test_speech_recognition_example.py ?

**Comment by Spycsh** [comment]
@emascarenhas Sure :) I've removed the redundant test.

**Comment by emascarenhas** [comment]
Looks good to me. @libinta , Please add run-test label.

File: optimum/habana/transformers/models/whisper/modeling_whisper.py [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+        elif past_key_values is not None:
+            past_key_values_length = past_key_values.get_seq_length()
+
+        if cache_position is None:
+            cache_position = torch.arange(
+                past_key_values_length, past_key_values_length + input_shape[1], device=inputs_embeds.device
+            )
+
+        if position_ids is None and token_idx:
+            position_ids = (token_idx - 1).unsqueeze(0)
```
Comment (regisss) [inline optimum/habana/transformers/models/whisper/modeling_whisper.py:316]: Shouldn't it be the following?
```suggestion
        if position_ids is None:
            if token_idx is not None:
                position_ids = (token_idx - 1).unsqueeze(0)
            else:
                position_ids = cache_position.unsqueeze(0)
```

File: optimum/habana/transformers/models/whisper/modeling_whisper.py [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+                        value_states = past_key_value.value_cache[self.layer_idx]
+                    else:
+                        past_key_value.key_cache.append(key_states)
+                        past_key_value.value_cache.append(value_states)
+                else:
+                    # save all key/value_states to cache to be re-used for fast auto-regressive generation
+                    cache_position = cache_position if not is_cross_attention else None
+                    # change cache_position to token_idx
+                    key_states, value_states = past_key_value.update(
+                        key_states, value_states, self.layer_idx, {"cache_position": token_idx}
```
[the remark this answers is not in the document]
Comment (Spycsh) [reply optimum/habana/transformers/models/whisper/modeling_whisper.py]: Yes you are right, and I have updated this.

File: optimum/habana/transformers/models/whisper/modeling_whisper.py (hunk 1 above)
Comment (Spycsh) [reply optimum/habana/transformers/models/whisper/modeling_whisper.py]: Thanks!

**Comment by Spycsh** [comment]
Hi @regisss , thanks for review:) I've updated the code based on your review.
